In [ ]:
import pandas as pd
from tqdm import tqdm
from sklearn.metrics import f1_score
import numpy as np
import explainer
import joblib

import warnings
warnings.filterwarnings("ignore")

In [ ]:
# KDD99 dataset
exclusive_rules = pd.read_csv('../Rules/KDD99_exclusive_rules.csv')
X_test = pd.read_csv('../Datasets/KDD99/X_test.csv')
y_test = pd.read_csv('../Datasets/KDD99/y_test.csv')
X_test['label'] = y_test
model = joblib.load('../Models/KDD99.pkl')
explainer.apply_exclusive_rules(X_test, exclusive_rules)
explainer.apply_exclusive_rules(X_test, exclusive_rules, model)

c:\Users\HP\AppData\Local\Programs\Python\Python311\Lib\pickle.py:1718: UserWarning: [22:12:26] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\data\../common/error_msg.h:82: If you are loading a serialized model (like pickle in Python, RDS in R) or
configuration generated by an older version of XGBoost, please export the model by calling
`Booster.save_model` from that version first, then load it back in current version. See:

    https://xgboost.readthedocs.io/en/stable/tutorials/saving_model.html

for more details about differences between saving model and serializing.

  setstate(state)
100%|██████████| 5889/5889 [00:32<00:00, 179.78it/s] 


Test set coverage: 69.76%
Test coverage F1-score: 1.00
--------------------------------------
Total Skipped: 8804 (30.24%)
 - Due to Conflicts: 23 (0.08%)
 - Due to No Rule Coverage: 8781 (30.16%)
--------------------------------------
Using Majority class to predict nonovered cases.
F1-score (if using entire dataset): 0.6916


100%|██████████| 5889/5889 [00:12<00:00, 479.88it/s] 


Test set coverage: 69.76%
Test coverage F1-score: 1.00
--------------------------------------
Total Skipped: 8804 (30.24%)
 - Due to Conflicts: 23 (0.08%)
 - Due to No Rule Coverage: 8781 (30.16%)
--------------------------------------
Using model predict non-covered cases.
F1-score (if using entire dataset): 0.9983


Resumee : 

SANS modèle:
- Couverture: 69.76%
- F1 sur couvert: 1.00 (parfait!)/ car sur notre zone de couverture il y a 100% de precision
- Non couverts: 30.24% (dont 0.08% conflits, 30.16% sans règle)
- F1 total (classe maj): 0.6916

AVEC modèle:
- Même couverture: 69.76%
- F1 sur couvert: 1.00
- F1 total (XGBoost): 0.9983 (Goooooood!)

Les règles exclusives sont parfaites là où elles s'appliquent. XGBoost est excellent pour compléter (F1=0.998)

et y a peu de conflits (0.08%) donc on conclu que les règles sont tres cohérentes

In [5]:
# BIG15
exclusive_rules = pd.read_csv('../Rules/BIG15_exclusive_rules.csv')
X_test = pd.read_csv('../Datasets/BIG15/X_test.csv')
y_test = pd.read_csv('../Datasets/BIG15/y_test.csv')
X_test['label'] = y_test
model = joblib.load('../Models/BIG15.pkl')
explainer.apply_exclusive_rules(X_test, exclusive_rules)
explainer.apply_exclusive_rules(X_test, exclusive_rules, model)

100%|██████████| 29262/29262 [00:08<00:00, 3312.62it/s]


Test set coverage: 93.72%
Test coverage F1-score: 0.99
--------------------------------------
Total Skipped: 1735 (6.28%)
 - Due to Conflicts: 184 (0.67%)
 - Due to No Rule Coverage: 1551 (5.62%)
--------------------------------------
Using Majority class to predict nonovered cases.
F1-score (if using entire dataset): 0.9482


100%|██████████| 29262/29262 [00:13<00:00, 2128.51it/s]

Test set coverage: 93.72%
Test coverage F1-score: 0.99
--------------------------------------
Total Skipped: 1735 (6.28%)
 - Due to Conflicts: 184 (0.67%)
 - Due to No Rule Coverage: 1551 (5.62%)
--------------------------------------
Using model predict non-covered cases.
F1-score (if using entire dataset): 0.9884


BIG15 a des règles très couvrantes (93.72% !)
Légèrement moins parfaites (F1=0.99 vs 1.00)
XGBoost améliore encore

In [6]:
# UNSW-NB15 dataset
exclusive_rules = pd.read_csv('../Rules/UNSWNB15_exclusive_rules.csv')
test = pd.read_csv('../Datasets/UNSW-NB15/processed/le_test.csv')
X_test = test.drop(['id','attack_cat'], axis=1)
model = joblib.load('../Models/UNSW-NB15.pkl')
explainer.apply_exclusive_rules(X_test, exclusive_rules)
explainer.apply_exclusive_rules(X_test, exclusive_rules, model)

100%|██████████| 9591/9591 [00:04<00:00, 1946.36it/s]


Test set coverage: 75.04%
Test coverage F1-score: 0.79
--------------------------------------
Total Skipped: 43768 (24.96%)
 - Due to Conflicts: 43369 (24.73%)
 - Due to No Rule Coverage: 399 (0.23%)
--------------------------------------
Using Majority class to predict nonovered cases.
F1-score (if using entire dataset): 0.6844


100%|██████████| 9591/9591 [00:07<00:00, 1340.08it/s]


Test set coverage: 75.04%
Test coverage F1-score: 0.79
--------------------------------------
Total Skipped: 43768 (24.96%)
 - Due to Conflicts: 43369 (24.73%)
 - Due to No Rule Coverage: 399 (0.23%)
--------------------------------------
Using model predict non-covered cases.
F1-score (if using entire dataset): 0.7743


    Probleme de detection : 
                                Énormément de conflits (24.73% !)
                                Les règles se contredisent souvent
                                F1 sur couvert médiocre (0.79)
                                Donc UNSW-NB15 est un dataset plus difficile

In [7]:
# DoH20
exclusive_rules = pd.read_csv('../Rules/DOH20_exclusive_rules.csv')
X_test = pd.read_csv('../Datasets/DoH20/X_test.csv')
y_test = pd.read_csv('../Datasets/DoH20/y_test.csv')
X_test['label'] = y_test
model = joblib.load('../Models/DoH20.pkl')
explainer.apply_exclusive_rules(X_test, exclusive_rules)
explainer.apply_exclusive_rules(X_test, exclusive_rules, model)

100%|██████████| 65744/65744 [00:18<00:00, 3597.18it/s]


Test set coverage: 98.90%
Test coverage F1-score: 1.00
--------------------------------------
Total Skipped: 593 (1.1%)
 - Due to Conflicts: 592 (1.1%)
 - Due to No Rule Coverage: 1 (0.0%)
--------------------------------------
Using Majority class to predict nonovered cases.
F1-score (if using entire dataset): 0.9896


100%|██████████| 65744/65744 [00:19<00:00, 3454.25it/s]

Test set coverage: 98.90%
Test coverage F1-score: 1.00
--------------------------------------
Total Skipped: 593 (1.1%)
 - Due to Conflicts: 592 (1.1%)
 - Due to No Rule Coverage: 1 (0.0%)
--------------------------------------
Using model predict non-covered cases.
F1-score (if using entire dataset): 0.9990


Ici DoH20 est "facile" : presque tout est couvert par des règles parfaites, Très peu de cas problématiques

In [9]:
# CIC17
exclusive_rules = pd.read_csv('../Rules/CIC2017_exclusive_rules.csv')
X_test = pd.read_parquet('../Datasets/CIC2017/X_test.parquet')
y_test = pd.read_csv('../Datasets/CIC2017/y_test.csv')
X_test['label'] = y_test
model = joblib.load('../Models/CIC17.pkl')
explainer.apply_exclusive_rules(X_test, exclusive_rules)
explainer.apply_exclusive_rules(X_test, exclusive_rules, model)

100%|██████████| 823366/823366 [17:12<00:00, 797.60it/s] 


Test set coverage: 92.63%
Test coverage F1-score: 0.99
--------------------------------------
Total Skipped: 37186 (7.37%)
 - Due to Conflicts: 3476 (0.69%)
 - Due to No Rule Coverage: 33710 (6.68%)
--------------------------------------
Using Majority class to predict nonovered cases.
F1-score (if using entire dataset): 0.9279


100%|██████████| 823366/823366 [19:56<00:00, 687.87it/s]  


Test set coverage: 92.63%
Test coverage F1-score: 0.99
--------------------------------------
Total Skipped: 37186 (7.37%)
 - Due to Conflicts: 3476 (0.69%)
 - Due to No Rule Coverage: 33710 (6.68%)
--------------------------------------
Using model predict non-covered cases.
F1-score (if using entire dataset): 0.9921


 le prlus grand:  c'est pourquoi il utilise .parquet (plus efficace que CSV)

 Les règles exclusives sont excellentes surtout quand elles existent
F1 de 0.99-1.00 sur les cas couverts pour 4/5 datasets

UNSW-NB15 est l'exception (dataset plus complexe) moins bon resultas 

La couverture varie énormément selon les datasets
DoH20 : 99% couvert → excellent pour l'explicabilité

KDD99 : 70% couvert → bon mais 30% (les sans regles) reste "boîte noire"

UNSW : beaucoup de conflits(24.73%) → règles moins fiables

3. Le modèle XGBoost est toujours utile
Améliore toujours le F1 total

Particulièrement important pour KDD99 (qui passe de 0.69 a 0.998)

Même pour UNSW (0.68 → 0.77)

4. Les conflits sont rares... sauf pour UNSW
UNSW a 24.73% de conflits ==> les règles se contredisent

Les autres datasets ont <1.1% de conflits